### PR TITLE
Extract API subpackage

### PR DESCRIPTION
**Pull request by thisac**
**Context:**
Due to the XCC integration, most of the Xanadu API subpackage is now gone.

**Description of the Change:**
* `strawberryfields.api.results.py` and `strawberryfields.api.devicespec.py` are moved to the root `strawberryfields` folder, and the `strawberryfields.api` folder is removed.
* The `FailedJobError` is moved from `strawberryfields.api.__init__.py` to `strawberryfields.engine.py` (where it's used).

**Benefits:**
Things look cleaner.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -247,6 +247,13 @@
   connection usages, and `DeviceSpec.target` is retrieved from the device specification rather than
   passed at initialization.
 
+* The `api` subpackage has been removed and the contained `DeviceSpec` and `Result` classes have
+  been moved to the root `strawberryfields` folder. They can now be imported as follows:
+
+  ```python
+  from strawberryfields import DeviceSpec, Result
```
Comment (Mandrenkov) [inline .github/CHANGELOG.md:254]: Is this the preferred access pattern? Or would we rather users write
```python
import strawberryfields as sf
# sf.DeviceSpec
# sf.Result
```

File: strawberryfields/engine.py [hunk 2]
```diff
@@ -39,6 +40,10 @@
 __all__ = ["BaseEngine", "LocalEngine", "BosonicEngine"]
 
 
+class FailedJobError(Exception):
+    """Raised when a job had a failure on the server side."""
```
Comment (Mandrenkov) [inline strawberryfields/engine.py:44]: I was wondering where this went. Good call! 👍

File: strawberryfields/__init__.py [hunk 3]
```diff
@@ -31,6 +31,8 @@
 from .tdm import TDMProgram
 from .plot import plot_wigner, plot_fock, plot_quad
 from . import tdm
+from .devicespec import DeviceSpec
+from .result import Result
```
Comment (Mandrenkov) [inline strawberryfields/__init__.py:35]: Should we add these to `__all__` as well?

File: tests/api/conftest.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """
-Test fixtures and shared functions for strawberryfields.api tests
+Test fixtures and shared functions for strawberryfields API tests
```
Comment (Mandrenkov) [inline tests/api/conftest.py:15]: Right?
```suggestion
Test fixtures and shared functions for Strawberry Fields API tests
```

File: tests/frontend/test_post_processing.py [hunk 5]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """
-Unit tests for strawberryfields.api.post_processing
+Unit tests for strawberryfields.utils.post_processing
```
Comment (Mandrenkov) [inline tests/frontend/test_post_processing.py:15]: Whoops! 🤭

**Review by Mandrenkov — APPROVED** [review]
Great work, Theo! 🧹

File: strawberryfields/__init__.py (hunk 3 above)
Comment (thisac) [reply strawberryfields/__init__.py:35]: I was thinking about that as well. Actually, now when you mention it, I think we probably should!

File: .github/CHANGELOG.md (hunk 1 above)
Comment (thisac) [reply .github/CHANGELOG.md:254]: True, I think your suggestion is better.

**Pull request merged**